### PR TITLE
executor, statistics: analyze with null count.

### PR DESCRIPTION
**Pull request by hanfei1991**
collect smaples for every single column.

### CODE REVIEW

File: statistics/histogram.go [hunk 1]
```diff
@@ -30,8 +30,9 @@ import (
 
 // Histogram represents statistics for a column or index.
 type Histogram struct {
-	ID  int64 // Column ID.
-	NDV int64 // Number of distinct values.
+	ID        int64 // Column ID.
+	NDV       int64 // Number of distinct values.
+	NullCount int64
```
Comment (shenli) [inline statistics/histogram.go]: Add comment for this attribute.

File: executor/analyze.go [hunk 2]
```diff
@@ -189,44 +185,58 @@ func analyzeIndex(exec *XSelectIndexExec) analyzeResult {
...
+
 // CollectSamplesAndEstimateNDVs collects sample from the result set using Reservoir Sampling algorithm,
 // and estimates NDVs using FM Sketch during the collecting process.
 // See https://en.wikipedia.org/wiki/Reservoir_sampling
 // Exported for test.
-func CollectSamplesAndEstimateNDVs(e ast.RecordSet, numCols int) (count int64, samples []*ast.Row, ndvs []int64, err error) {
-	var sketches []*statistics.FMSketch
-	for i := 0; i < numCols; i++ {
-		sketches = append(sketches, statistics.NewFMSketch(maxSketchSize))
+func CollectSamplesAndEstimateNDVs(e ast.RecordSet, numCols int) ([]*SampleCollector, error) {
```
Comment (shenli) [inline executor/analyze.go]: Add comment for the first return value.

File: executor/analyze.go [hunk 3]
```diff
@@ -189,44 +185,58 @@ func analyzeIndex(exec *XSelectIndexExec) analyzeResult {
...
 
+// SampleCollector will collect samples and calculate the count and ndv of an attribute.
+type SampleCollector struct {
+	samples   []types.Datum
+	NullCount int64
+	Count     int64
+	Sketch    *statistics.FMSketch
+}
+
+func (c *SampleCollector) insert(d types.Datum) error {
```
Comment (shenli) [inline executor/analyze.go]: How about rename `insert` to `collect`?

**Comment by hanfei1991** [comment]
@shenli PTAL

**Comment by shenli** [comment]
LGTM
@zimulala @XuHuaiyu PTAL

File: executor/analyze.go [hunk 4]
```diff
@@ -189,44 +185,59 @@ func analyzeIndex(exec *XSelectIndexExec) analyzeResult {
...
+func (c *SampleCollector) collect(d types.Datum) error {
+	if d.IsNull() {
+		c.NullCount++
+		return nil
+	}
+	c.Count++
+	if len(c.samples) < maxSampleCount {
+		c.samples = append(c.samples, d)
+	} else {
+		shouldAdd := rand.Int63n(c.Count-c.NullCount) < maxSampleCount
```
Comment (zimulala) [inline executor/analyze.go]: Use `c.Count` instead of `c.Count-c.NullCount`?

**Comment by zimulala** [comment]
LGTM

**Pull request merged**